api/workload/v2/workload.go

```diff
@@ -0,0 +1,160 @@
+//
+// See https://github.com/spiffe/spiffe/blob/master/standards/X509-SVID.md
+type X509SVID struct {
+	SPIFFEID     string
+	PrivateKey   crypto.Signer
+	Certificates []*x509.Certificate
+	TrustBundle  *x509.CertPool
+}
+
+// WorkloadIdentityWatcher is implemented by consumers who wish to be updated on SVID changes.
+type WorkloadIdentityWatcher interface {
+	// UpdateX509SVIDs indicates to the Watcher that the SVID has been updated
+	UpdateX509SVIDs(*X509SVIDs)
+
+	// OnError indicates an error occurred.
+	OnError(err error)
+
+	// OnConnection indicates a change in the connection state to the SPIFFE agent.
+	OnConnection(connected bool)
+}
+
+// Client interacts with the SPIFFE Workload API.
+type Client struct {
+	logger         *logrus.Logger
+	watcher        WorkloadIdentityWatcher
+	addr           string
+	wg             sync.WaitGroup
+	connectionChan chan bool
+	reader         *streamReader
+	ctx            context.Context
+	cancelFn       func()
+	streamManager  *streamManager
+}
+
+// Option configures the workload client.
+type Option func(*Client)
+
+// Addr specifies the unix socket address of the SPIFFE agent.
+func Addr(addr string) Option {
+	return func(w *Client) {
+		w.addr = addr
+	}
+}
+
+// Logger specifies the logger to use.
+func Logger(logger *logrus.Logger) Option {
+	return func(w *Client) {
+		w.logger = logger
+	}
+}
+
+// NewClient returns a new Workload API client.
+func NewClient(watcher WorkloadIdentityWatcher, opts ...Option) (*Client, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	w := &Client{
+		logger:         logrus.StandardLogger(),
+		addr:           DefaultAgentAddress,
+		watcher:        watcher,
+		connectionChan: make(chan bool, 1),
+		ctx:            ctx,
+		cancelFn:       cancel,
+	}
+	for _, opt := range opts {
+		opt(w)
+	}
+	w.streamManager = newStreamManager(ctx, w.logger, w.addr, w.connectionChan)
+	w.reader = newStreamReader(ctx, w.logger, w.streamManager)
+	return w, nil
+}
+
+// Start starts the client.
+//
+// This blocks on setting up an initial connection to a SPIFFE Workload API.
+// The passed context may be used to set a deadline for this setup.
+func (c *Client) Start(ctx context.Context) error {
+	c.wg.Add(1)
+	if err := c.streamManager.Start(ctx); err != nil {
+		c.cancelFn()
+		return err
+	}
+	go c.run(c.ctx)
+	return nil
+}
+
+// Stop stops the client.
+func (c *Client) Stop() error {
+	c.cancelFn()
+	c.wg.Wait()
+	return nil
+}
+
+func (c *Client) run(ctx context.Context) {
+	defer c.wg.Done()
+
+	c.handleUpdates(ctx)
+
+	c.reader.Stop()
+	if c.connectionChan != nil {
+		close(c.connectionChan)
+		c.logger.WithField("queued", len(c.connectionChan)).Debug("Emptying connection chan.")
+		for range c.connectionChan {
+		}
+		c.connectionChan = nil
+	}
+}
+
+func (c *Client) handleUpdates(ctx context.Context) {
+	for {
+		select {
+		case resp, ok := <-c.reader.Chan:
+			if ok {
+				c.onReceive(resp)
+			}
```

Comment:
moving the stream handling logic here would mean you need to handle the !ok case. This would mean the reader shut itself down due to a connection error and you would want to send a connection event and then redial.